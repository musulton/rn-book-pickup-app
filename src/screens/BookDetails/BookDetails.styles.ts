import {ImageStyle, StyleSheet, TextStyle, ViewStyle} from "react-native";
import {COLORS, SIZES} from "../../constants";

interface Styles {
    container: ViewStyle
    imageWrapper: ViewStyle
    scrollContainer: ViewStyle
    image: ImageStyle
    textWrapper: TextStyle
    title: TextStyle
    author: TextStyle
    edition: TextStyle
    description: TextStyle
    buttonWrapper: ViewStyle
    descriptionLabel: TextStyle
}

const styles = StyleSheet.create<Styles>({
    container: {
        display: "flex",
        flex: 1,
        backgroundColor: COLORS.background,
        height: "100%"
    },
    scrollContainer: {
    },
    imageWrapper: {
        justifyContent: "center",
        alignItems: "center",
        padding: 20,
    },
    image: {
        width: 150,
        height: 200,
        resizeMode: "stretch",
        borderRadius: SIZES.borderRadius
    },
    textWrapper: {
        margin: SIZES.padding1
    },
    title: {
        fontSize: 22,
        fontWeight: "500"
    },
    author: {
        fontSize: 14,
        color: COLORS.lightGray,
        marginVertical: 5
    },
    edition: {
        fontSize: 18,
        fontWeight: "500"
    },
    descriptionLabel: {
        fontSize: 14,
        lineHeight: 20,
        paddingHorizontal: SIZES.padding1,
        fontWeight: "bold"
    },
    description: {
        fontSize: 14,
        color: COLORS.lightGray,
        lineHeight: 20,
        paddingHorizontal: SIZES.padding1
    },
    buttonWrapper: {
        padding: 20,
        display: "flex",
        flexDirection: "row",
        justifyContent: "center"
    }
});

export default styles
