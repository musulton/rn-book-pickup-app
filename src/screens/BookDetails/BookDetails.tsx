import {Text, View} from "react-native";
import React from "react";

import styles from "./BookDetails.styles";

const BookDetails: React.FC = () => {
    return (
        <View style={styles.container}>
            <Text>Book detail screen</Text>
        </View>
    );
}

export default BookDetails
