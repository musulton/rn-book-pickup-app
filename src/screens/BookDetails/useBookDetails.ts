import React from "react";

interface UseBookDetails {
    getter: {
        selectedDate: Date
        datePickerVisible: boolean
    }
    setter: {
        selectedDate: (date: Date) => void
        datePickerVisible: (visibility: boolean) => void
    }
}

const useBookDetails: () => UseBookDetails = () => {
    const [selectedDate, setSelectedDate] = React.useState(new Date());
    const [datePickerVisible, setDatePickerVisible] = React.useState(false);

    return {
        getter: { selectedDate, datePickerVisible },
        setter: { selectedDate: setSelectedDate, datePickerVisible: setDatePickerVisible }
    }
}

export default useBookDetails
