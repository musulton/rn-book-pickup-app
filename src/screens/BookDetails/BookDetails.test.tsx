import {useDispatch} from "react-redux";
import {Alert} from "react-native";
import {cleanup, fireEvent, render, screen} from "@testing-library/react-native";

import {book} from "../../fixtures/testData";

import BookDetails from "./BookDetails";

jest
    .mock("../../utils/function", () => ({
        getRandomColor: jest.fn(() => "white")
    }))
    .mock("react-redux")

describe("<BookDetails />", () => {
    const mockBook = {
        ...book,
        coverImg: "image.jpg"
    }
    const mockDate = new Date(2023, 3, 1)
    const mockProps = {
        navigation: {
            goBack: jest.fn()
        },
        route: {
            params: {
                book: mockBook
            }
        }
    }

    afterEach(() => {
        cleanup()
    })

    beforeAll(() => {
        jest.useFakeTimers("modern");
        jest.setSystemTime(mockDate);
    });

    afterAll(() => {
        jest.useRealTimers();
    });

    it("should render correctly", () => {
        render(<BookDetails {...mockProps} />)

        expect(screen.toJSON()).toMatchSnapshot()
    })

    it("should display pickup date modal when button is pressed", async () => {
        render(<BookDetails {...mockProps} />)

        const pickupButton = await screen.findByText("Set Pick-up Date")
        fireEvent.press(pickupButton)

        const bookNowButton = await screen.findByText("Book Now")
        const cancelButton = await screen.findByText("Cancel")

        expect(bookNowButton).toBeTruthy()
        expect(cancelButton).toBeTruthy()
    })

    it("should call onSubmit when book now button is pressed", async () => {
        jest.spyOn(Alert, "alert")
        const mockDispatch = jest.fn()
        useDispatch.mockReturnValue(mockDispatch)

        render(<BookDetails {...mockProps} />)

        const pickupButton = await screen.findByText("Set Pick-up Date")
        fireEvent.press(pickupButton)

        const bookNowButton = await screen.findByText("Book Now")
        fireEvent.press(bookNowButton)

        expect(mockDispatch).toHaveBeenCalledWith({
            type: "books/add",
            payload: {
                ...mockBook,
                pickupDate: mockDate.toLocaleDateString()
            }
        })
        expect(mockProps.navigation.goBack).toHaveBeenCalled()
        expect(Alert.alert).toHaveBeenCalledWith("Hooray", "The book has been successfully booked")
    })
})
