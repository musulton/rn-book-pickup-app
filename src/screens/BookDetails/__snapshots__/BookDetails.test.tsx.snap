// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookDetails /> should render correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#FEFEFE",
      "display": "flex",
      "flex": 1,
      "height": "100%",
    }
  }
>
  <RCTScrollView
    contentContainerStyle={{}}
  >
    <View>
      <View
        style={
          [
            {
              "alignItems": "center",
              "justifyContent": "center",
              "padding": 20,
            },
            {
              "backgroundColor": "white",
            },
          ]
        }
      >
        <Image
          source={
            {
              "uri": "image.jpg",
            }
          }
          style={
            {
              "borderRadius": 6,
              "height": 200,
              "resizeMode": "stretch",
              "width": 150,
            }
          }
        />
      </View>
      <View
        style={
          {
            "margin": 20,
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 22,
              "fontWeight": "500",
            }
          }
        >
          title
        </Text>
        <Text
          style={
            {
              "color": "#64676D",
              "fontSize": 14,
              "marginVertical": 5,
            }
          }
        >
          author
        </Text>
        <Text
          style={
            {
              "fontSize": 18,
              "fontWeight": "500",
            }
          }
        >
          edition
        </Text>
      </View>
      <Text
        style={
          {
            "fontSize": 14,
            "fontWeight": "bold",
            "lineHeight": 20,
            "paddingHorizontal": 20,
          }
        }
      >
        Description
      </Text>
      <Text
        style={
          {
            "color": "#64676D",
            "fontSize": 14,
            "lineHeight": 20,
            "paddingHorizontal": 20,
          }
        }
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
      </Text>
    </View>
  </RCTScrollView>
  <View
    style={
      {
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "center",
        "padding": 20,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#039BE5",
          "borderRadius": 6,
          "display": "flex",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "#FFFFFF",
            "fontSize": 16,
            "fontWeight": "500",
            "padding": 20,
          }
        }
      >
        Set Pick-up Date
      </Text>
    </View>
    <Modal
      animationType="none"
      hardwareAccelerated={false}
      isVisible={false}
      onHide={[Function]}
      transparent={true}
      visible={false}
    />
  </View>
</View>
`;
